### PR TITLE
container/controller: avoid cancellation with forked pull context

### PR DESCRIPTION
**Pull request by stevvooe**
Context cancellations were previously causing `Prepare` to fail
completely on re-entrant calls. To prevent this, we filtered out cancels
and deadline errors. While this allowed the service to proceed without
errors, it had the possibility of interrupting long pulls, causing the
pull to happen twice.

This PR forks the context of the pull to match the lifetime of
`Controller`, ensuring that for each task, the pull is only performed
once. It also ensures that multiple calls to `Prepare` are re-entrant,
ensuring that the pull resumes from its original position.

Signed-off-by: Stephen J Day stephen.day@docker.com

Further mitigation to #24447.

### CODE REVIEW

**Comment by stevvooe** [comment]
cc @cpuguy83 @tiborvass @aluzzardi @aaronlehmann

File: daemon/cluster/executor/container/controller.go [hunk 1]
```diff
@@ -86,24 +92,41 @@ func (r *controller) Prepare(ctx context.Context) error {
...
+			// cancelled on updates.
+			var pctx context.Context
+
+			r.pulled = make(chan struct{})
+			pctx, r.cancelPull = context.WithCancel(context.Background()) // TODO(stevvooe): Bind a context to the entire controller.
+
+			go func() {
+				defer close(r.pulled)
+				err := r.adapter.pullImage(pctx)
+				r.pullErrOnce.Do(func() { r.pullErr = err })
```
Comment (aaronlehmann) [inline daemon/cluster/executor/container/controller.go]: I don't understand why we need this `sync.Once`. How can we invoke this goroutine multiple times?

Comment (stevvooe) [inline daemon/cluster/executor/container/controller.go]: It provides the barrier for the error value. Do you think `r.pulled` is sufficient?

**Comment by stevvooe** [comment]
Just gave this a full test and we are looking solid.

File: daemon/cluster/executor/container/controller.go (hunk 1 above)
Comment (aaronlehmann) [inline daemon/cluster/executor/container/controller.go]: Yeah, I thought the controller methods were only called one at a time, so checking `r.pulled` would be enough. Did I misunderstand?

Comment (aaronlehmann) [inline daemon/cluster/executor/container/controller.go]: In terms of a barrier, a channel receive should act as a memory barrier, so yeah, I think we'd be okay without the `sync.Once`.

**Comment by aaronlehmann** [comment]
A test timed out.

```
04:36:47 --- FAIL: Test (1187.00s)
04:36:47 panic: DockerSuite.TestRunTmpfsMountsWithOptions test timed out after 5m0s [recovered]
04:36:47 	panic: DockerSuite.TestRunTmpfsMountsWithOptions test timed out after 5m0s
```

Unrelated?

**Comment by stevvooe** [comment]
> Unrelated?

Completely, restarted builder.

**Comment by aaronlehmann** [comment]
LGTM

**Comment by aluzzardi** [comment]
LGTM

**Pull request merged**